src/java/io/bazel/rulesscala/test_discovery/ArchiveEntries.scala

```diff
@@ -0,0 +1,36 @@
+package io.bazel.rulesscala.test_discovery
+
+import java.io.{File, FileInputStream}
+import java.util.jar.{JarEntry, JarInputStream}
+
+object ArchiveEntries {
+  def listClassFiles(file: File): Stream[String] = {
+    val allEntries = if (file.isDirectory)
+      directoryEntries(file).map(_.stripPrefix(file.toString).stripPrefix("/"))
+    else
+      jarEntries(new JarInputStream(new FileInputStream(file)))
+
+    allEntries.filter(_.endsWith(".class"))
+  }
+
+  private def getJarEntryOrCloseStream(jarInputStream: JarInputStream): Option[JarEntry] = {
+    val entry = Option(jarInputStream.getNextJarEntry)
+
+    if (entry.isEmpty)
+      jarInputStream.close()
+
+    entry
+  }
+
+  private def jarEntries(jarInputStream: JarInputStream): Stream[String] =
+    Stream.continually(getJarEntryOrCloseStream(jarInputStream))
```

Comment:
Maybe instead of `getJarEntryOrCloseStream` wrap whole stream in `try` and close stream in `finally`?

Reply:
Is that possible with a lazy stream? Or am I missing the point in what you're saying?

Reply:
Right `lazy`... Then please ignore this comment. However I think this is a bit dangerous. What if stream is not consumed till the end, but maybe this is not a problem in this case.

Reply:
Yes, it's a risk, but I think impact is small. Current code also won't close in case exception is thrown before the stream is closed